pkg/virtual/generic/cmd/cmd.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	genericapiserveroptions "k8s.io/apiserver/pkg/server/options"
+	"k8s.io/client-go/pkg/version"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog/v2"
+	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/templates"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	kubeoptions "k8s.io/kubernetes/pkg/kubeapiserver/options"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	virtualbuilders "github.com/kcp-dev/kcp/pkg/virtual/generic/builders"
+	virtualrootapiserver "github.com/kcp-dev/kcp/pkg/virtual/generic/rootapiserver"
+)
+
+type APIServerOptions struct {
+	Output io.Writer
+
+	SecureServing     *genericapiserveroptions.SecureServingOptionsWithLoopback
+	Authentication    *genericapiserveroptions.DelegatingAuthenticationOptions
+	SubCommandOptions SubCommandOptions
+}
+
+type SubCommandDescription struct {
+	Name  string
+	Use   string
+	Short string
+	Long  string
+}
+
+type SubCommandOptions interface {
+	Description() SubCommandDescription
+	AddFlags(flags *pflag.FlagSet)
+	Validate() []error
+	InitializeBuilders() ([]virtualrootapiserver.InformerStart, []virtualbuilders.VirtualWorkspaceBuilder, error)
```

Comment:
nit: Initialize usually is some func with side-effect. This is more a constructor, i.e. NewBuilders or CreateBuilders makes sense.

Reply:
Renamed the method to `PrepareVirtualWorkspaces` since the `VirtualWorkspaceBuilder` type has been changed to the `VirtualWorkspace` interface